libHguX/HGU_XLut.3
'\" t
.TH HGU_XLut 3 "$Id$" "MRC HGU" "RICHARD\'S MOTIF-LIBRARY"
.SH NAME
HGU_XLut, HGU_XSetLut, HGU_XGtransSetLut, HGU_XGtransSetGreys,
HGU_XGtransSetOverlay, HGU_XGtransSetDefaultLut
\- routines to manipulate X11 window colourmaps with respect to
the grey-transfer function of a PframeX structure.
.SH SYNOPSIS
These are convenience routines usually used in conjuction with the
woolz display procedure picframeX(3) and provide a means of setting
X11 window colourmaps compatible with the grey-level transfer function.
.nf
.sp
#include <X11/Xlib.h>
#include <disp_utilsX.h>

int HGU_XSetLut( dpy, win, cols, ncols )
Display		*dpy;
Window		win;
XColor		*cols;
int		ncols;

int HGU_XGtransSetLut( dpy, win, colormap, gtrans )
Display		*dpy;
Window		win;
unsigned char	colormap[3][256];
GTrans		*gtrans;

int HGU_XGtransSetGreys( dpy, win, colormap, planes, gtrans )
Display		*dpy;
Window		win;
unsigned char	colormap[3][256];
unsigned long	planes;
GTrans		*gtrans;

int HGU_XGtransSetOverlay( dpy, win, plane, r, g, b, gtrans )
Display		*dpy;
Window		win;
unsigned long	plane;
int		r, g, b;
GTrans		*gtrans;

int HGU_XGtransSetDefaultLut( dpy, win, gtrans )
Display		*dpy;
Window		win;
GTrans		*gtrans;

.fi
.SH DESCRIPTION
.LP
\fBHGU_XSetLut()\fR - set as many of the colours from the cols array
as possible in the colormap of the given window. All allocated colorcells
will be freed which means that other applications could change the same
colormap entries.
.LP
\fBHGU_XGtransSetLut()\fR - set the window colormap to correspond to the
given colormap and grey-level transform. The function calculates the
colorcells that need changing and then calls \fBHGU_XSetLut\fR. Images
displayed using \fBpicframeX()\fR will appear as if displayed using the
given colormap even though the actual display values have been restricted
to a smaller range. The colormap values are unsigned byte and are
multiplied by 256 to convert to X11 values which are 16-bit. The colour
values are transferred as implied by the grey-transfer function which will
not in general respect overlay plane settings. If an overlay plane is
required then use the procedures \fBHGU_XGtransSetGreys\fR and
\fBHGU_XGtransSetOverlay\fR to set the grey-value colours and overlay plane
respectively.
.LP
\fBHGU_XGtransSetGreys()\fR - set the grey-level colours from
\fBcolormap\fR for all pixel values which do not set bits in the planes
bitmask \fBplane\fR. The values are transferred as implied by \fBgtrans\fR.
.LP
\fBHGU_XGtransSetOverlay()\fR - set the colour of an overlay plane
for values implied by the \fBgtrans\fR structure.
.LP
.SH FILES
.nf
~richard/Motif/include           include directory
~richard/Motif/lib/              library directory
~richard/Motif/lib/librichXm.a   library archive file
.fi
.SH "SEE ALSO"
X11 & Motif documentation

.SH BUGS

.SH COPYRIGHT
Copyright 1993, Medical Research Council, UK.
.SH AUTHOR
Richard Baldock, MRC Human Genetics Unit, Edinburgh, UK.



libHguX/HGU_XLut.c
#if defined(__GNUC__)
#ident "MRC HGU $Id:"
#else
#if defined(__SUNPRO_C) || defined(__SUNPRO_CC)
#pragma ident "MRC HGU $Id:"
#else static char _HGU_XLut_c[] = "MRC HGU $Id:";
#endif
#endif
/*!
* \file         HGU_XLut.c
* \author       Richard Baldock <Richard.Baldock@hgu.mrc.ac.uk>
* \date         Wed Apr 29 08:18:22 2009
* \version      MRC HGU $Id$
*               $Revision$
*               $Name$
* \par Address:
*               MRC Human Genetics Unit,
*               Western General Hospital,
*               Edinburgh, EH4 2XU, UK.
* \par Copyright:
* Copyright (C) 2005 Medical research Council, UK.
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be
* useful but WITHOUT ANY WARRANTY; without even the implied
* warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
* PURPOSE.  See the GNU General Public License for more
* details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the Free
* Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA  02110-1301, USA.
* \ingroup      HGU_X
* \brief         Convenience routines to manipulate look-up tables
*               using the grey-transform associated with a PframeX structure
* \todo         -
* \bug          None known
*
* Maintenance log with most recent changes at top of list.
*/

#include <stdio.h>
#include <stdlib.h>

#include <X11/Xlib.h>

#include <HGU_XUtils.h>

static int	       	ncells;
static unsigned long	pix_vals[256];
static int	       	LUT_allocated = 0;

/* --------------------- HGU_XAllocLut ----------------------------
 *
 *	Allocates but does not set colour cells.
 *	Sets local globals pix_vals[] and ncells.
 * ---------------------------------------------------------------- */

int HGU_XAllocLut (Display     	*dpy,
		   Window      	win)
{
  unsigned long		plane_masks[1];
  XWindowAttributes	win_att;

  if (LUT_allocated == 1)
    return (0);			/* is this an error ?? */

  /* set up the window attributes  */
  if (XGetWindowAttributes (dpy, win, &win_att) == 0 )
    {
      fprintf (stderr, "HGU_XAllocLut: failed to get window attributes\n");
      exit (1);
    }

  /* check for monochrome */
  if (win_att.depth == 1)
    return (0);

  /* if the colormap seems funny, put in the default */
  if (!win_att.colormap)
    win_att.colormap = DefaultColormap (dpy, DefaultScreen (dpy));

  /* get ownership of all remaining colorcells */
  ncells = 0;
  while (XAllocColorCells (dpy, win_att.colormap,
			   False, plane_masks, 0, &pix_vals[ncells], 1))
    {
      ncells++;
      if (ncells > 255)
	break;
    }

  LUT_allocated = 1;

  return (0);
}

/* ---------------------- HGU_XStoreLut -------------------------------
 *
 *	Stores given colour map in previously allocated colour cells
 *	if possible.
 * -------------------------------------------------------------------- */

int HGU_XStoreLut (Display     	*dpy,
		   Window      	win,
		   XColor      	*cols,
		   int		ncols)
{
    XWindowAttributes	win_att;
    XColor		new_cols[256];
    int			i, j, k;

  if (LUT_allocated == 0)
    {
      fprintf (stderr, "HGU_XStoreLut: No LUT allocated\n");
      return (1);
    }

  /* set up the window attributes  */
  if (XGetWindowAttributes (dpy, win, &win_att) == 0 )
    {
      fprintf (stderr, "HGU_XStoreLut: failed to get window attributes\n");
      exit (1);
    }

    /* discard colors we can't set */
    for (i = 0, k = 0; i < ncols; i++)
      {
	for (j = 0; j < ncells; j++)
	  {
	    if (cols[i].pixel == pix_vals[j])
	      {
		new_cols[k] = cols[i];
		k++;
		break;
	      }
	  }
      }

    /* set the colours */
    if (k > 0)
	XStoreColors (dpy, win_att.colormap, new_cols, k);

    XFlush (dpy);

    return (0);
}

/* --------------- HGU_XFreeLut -----------------------------------
 *
 *	Frees previously allocated colour cells.
 * ---------------------------------------------------------------- */

int	HGU_XFreeLut (Display      *dpy,
		      Window       win)
{
  XWindowAttributes	win_att;

  if (LUT_allocated == 0)
    {
      fprintf (stderr, "HGU_XFreeLut: No LUT allocated\n");
      return (1);
    }

  /* set up the window attributes  */
  if (XGetWindowAttributes (dpy, win, &win_att) == 0 )
    {
      fprintf (stderr, "HGU_XFreeLut: failed to get window attributes\n");
      exit (1);
    }

  /* free the colorcells */
      if (ncells > 0)
     	XFreeColors (dpy, win_att.colormap, pix_vals, ncells, (long) 0);
     
    XFlush (dpy);

    LUT_allocated = 0;

    return (0);
}

/* --------------- HGU_XSetLut -----------------------------------
 *
 *	Allocates, stores and frees look up table.
 * ---------------------------------------------------------------- */

int	HGU_XSetLut (Display     	*dpy,
		     Window      	win,
		     XColor      	*cols,
		     int		ncols)
{
  if ((HGU_XAllocLut (dpy, win) == 0)
    && (HGU_XStoreLut (dpy, win, cols, ncols) == 0)
      && (HGU_XFreeLut (dpy, win) == 0))
    {
      return (0);
    }
  else
    {
      return (1);
    }
}

/* --------------- HGU_XGtransSetLut -------------------------------- */

int HGU_XGtransSetLut(
Display		*dpy,
Window		win,
unsigned char	colormap[3][256],
GTrans		*gtrans)
{
    unsigned long	planes=0;

    return( HGU_XGtransSetGreys( dpy, win, colormap, planes, gtrans ) );
}

/* -------------- HGU_XGtransToXCols ---------------------------------
 *
 *	Transforms colour map into Lut.
 *	Pointer to return Lut and number of colours given as parameters.
 * -------------------------------------------------------------------- */

int	HGU_XGtransToXCols (Display		*dpy,
			    Window		win,
			    unsigned char	colormap[3][256],
			    unsigned long	planes,
			    GTrans		*gtrans,
			    XColor		*cols,	/* Return */
			    int			*ncols)	/* Return */
{
    int			g, G;
    XWindowAttributes	win_att;

    /* set up the window attributes  */
    if (XGetWindowAttributes (dpy, win, &win_att) == 0)
      {
	fprintf (stderr, 
		 "HGU_XGtransToXCols: failed to get window attributes\n");
	exit (1);
      }

    for (G = gtrans->basic.Gmin, *ncols = 0; G <= gtrans->basic.Gmax; G++)
      {
	g = (*gtrans->basic.transf_func)( G, (caddr_t) gtrans, 1 );
	g = (g > 255) ? 255 : g;
	g = (g < 0) ? 0 : g;

	if (g & planes)
	    continue;

	switch (win_att.depth)
	  {
	  case 1:
	    return (1);
	  case 4:
	    cols->pixel = G / 16;		/* cols[ncols].pixel */
	    break;
	  case 24:
	    cols->pixel = G + (G << 8) + (G << 16);
	    break;
	  case 8:
	  default:
	    cols->pixel = G;
	    break;
	  }

	cols->flags = DoRed | DoGreen | DoBlue;
	cols->red   = (short) colormap[0][g] * 256;
	cols->green = (short) colormap[1][g] * 256;
	cols->blue  = (short) colormap[2][g] * 256;

	(*ncols)++;
	cols++;

	if (*ncols >= 256)
	    break;
    }

    return (0);
}

/* -------------- HGU_XGtransSetGreys ---------------------------------
 *
 *	Genetates Lut and then sets it
 * -------------------------------------------------------------------- */

int	HGU_XGtransSetGreys (Display		*dpy,
			     Window		win,
			     unsigned char	colormap[3][256],
			     unsigned long	planes,
			     GTrans		*gtrans)
{
  XColor	cols[256];
  int		ncols;

  if ((HGU_XGtransToXCols (dpy, win, colormap, planes, gtrans,
			   cols, &ncols) == 0)
      && (HGU_XSetLut (dpy, win, cols, ncols) == 0))
    {
      return (0);
    }
  else
    {
      return (1);
    }
}

/* -------------- HGU_XGtransSetOverlay ------------------------------ */

int HGU_XGtransSetOverlay(
Display		*dpy,
Window		win,
unsigned long	plane,
int		r,
int		g,
int		b,
GTrans		*gtrans)
{
    XColor		cols[256];
    int			G, ncols;

    for(G = gtrans->basic.Gmin, ncols=0; G <= gtrans->basic.Gmax; G++){

	if( !(G & plane) )
	    continue;

	cols[ncols].pixel = G;
	cols[ncols].flags = DoRed | DoGreen | DoBlue;
	cols[ncols].red   = (short) r * 256;
	cols[ncols].green = (short) g * 256;
	cols[ncols].blue  = (short) b * 256;

	ncols++;
	if( ncols >= 256 )
	    break;
    }

    HGU_XSetLut( dpy, win, cols, ncols );
    return( 0 );
}

int HGU_XGtransSetDefaultLut(
Display		*dpy,
Window		win,
GTrans		*gtrans)
{
    unsigned char	colormap[3][256];
    int			i;

    for(i=0; i < 256; i++){
	colormap[0][i] = 255 - i;
	colormap[1][i] = 255 - i;
	colormap[2][i] = 255 - i;
    }

    HGU_XGtransSetLut( dpy, win, colormap, gtrans );
    return( 0 );
}
